File: epi_judge_python_solutions/int_as_list_add.py
from list_node import ListNode
from test_framework import generic_test


def add_two_numbers(L1, L2):

    place_iter = dummy_head = ListNode()
    carry = 0
    while L1 or L2 or carry:
        val = carry + (L1.data if L1 else 0) + (L2.data if L2 else 0)
        L1 = L1.next if L1 else None
        L2 = L2.next if L2 else None
        place_iter.next = ListNode(val % 10)
        carry, place_iter = val // 10, place_iter.next
    return dummy_head.next


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("int_as_list_add.py",
                                       'int_as_list_add.tsv', add_two_numbers))

File: epi_judge_cpp_solutions/int_as_list_add.cc
#include "list_node.h"
#include "test_framework/generic_test.h"

shared_ptr<ListNode<int>> AddTwoNumbers(shared_ptr<ListNode<int>> L1,
                                        shared_ptr<ListNode<int>> L2) {
  shared_ptr<ListNode<int>> dummy_head(new ListNode<int>);
  auto place_iter = dummy_head;
  int carry = 0;
  while (L1 || L2 || carry) {
    int val = carry + (L1 ? L1->data : 0) + (L2 ? L2->data : 0);
    L1 = L1 ? L1->next : nullptr;
    L2 = L2 ? L2->next : nullptr;
    place_iter->next = make_shared<ListNode<int>>(val % 10, nullptr);
    carry = val / 10, place_iter = place_iter->next;
  }
  return dummy_head->next;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"L1", "L2"};
  return GenericTestMain(args, "int_as_list_add.cc", "int_as_list_add.tsv",
                         &AddTwoNumbers, DefaultComparator{}, param_names);
}
